### ISSUE TITLE
Unable to use constraints with UNIQUE operator 

### ISSUE THREAD

**Issue by rodrickbrown** [mesosphere/marathon#2588]
I have 3 dedicated servers for Spark Mesos ELT jobs and another 6 servers dedicated to microservices I want to run the spark-mesos-shuffle-service on only the 3 dedicated spark slaves. 

On my 3 spark slaves attributes are set to 
rack:spark;zone:us-west-2c;machine:m3.2xlarge

on my 6 microservices slaves my attributes are set to 
rack:ms;zone:us-west-2a;machine:m4.4xlarge

The spark-mesos-shuffle-service binds to a fixed port 7337. I can't change this and I would like to have it start up only on my 3 spark servers and on a unique hosts. 

Basically on a successful deploy I should see 
spark-slave-1 shuffle-service
spark-slave-2 shuffle-service
spark-slave-3 shuffle-service

I've tried every combination of constraints I could think of for making this work and it seems impossible. 

Why won't these constraints work for the following attributes I have listed above? 

"constraints": [
    [  
      "hostname",
      "UNIQUE",
      "spark"
     ]

I end up with services starting on nodes with the following attributes 
rack:ms;zone:us-west-2a;machine:m4.4xlarge

**Comment by junneyang** [comment]
i think the constraints defination should like this:
"constraints": [
["hostname","UNIQUE"],
["hostname","CLUSTER","spark"],
]

**Comment by rodrickbrown** [comment]
@junneyang this doesn't work job is stuck deploying... 

$ cat spark-shuffle-service.marathon.json
{
    "id": "/mesos/spark-shuffle-service",
    "cmd": ". /opt/spark-1.5.1/sbin/spark-config.sh && . /opt/spark-1.5.1/bin/load-spark-env.sh && env && /usr/java/jdk1.8.0_60/jre/bin/java -cp /opt/spark-1.5.1/sbin/../conf/:/opt/spark-1.5.1/assembly/target/scala-2.11/spark-assembly-1.5.1-hadoop2.5.2.jar:/opt/spark-1.5.1/lib_managed/jars/datanucleus-core-3.2.10.jar:/opt/spark-1.5.1/lib_managed/jars/datanucleus-rdbms-3.2.9.jar:/opt/spark-1.5.1/lib_managed/jars/datanucleus-api-jdo-3.2.6.jar -Xms1g -Xmx1g org.apache.spark.deploy.mesos.MesosExternalShuffleService 1",
    "cpus": 0.5,
    "mem": 1024,
    "disk": 50,
    "user": "mesos",
    "instances": 3,
    "constraints": [
      [
        "hostname", "UNIQUE"
      ],
      [
        "hostname","CLUSTER","spark"
      ]
    ],

    "maxLaunchDelaySeconds": 3,
    "backoffFactor": 1.20,
		"upgradeStrategy": {
      "minimumHealthCapacity": 0.5,
      "maximumOverCapacity": 0.5
    }
}

**Comment by aquamatthias** [comment]
Hey @rodrickbrown you use hostname constraints, but it looks like you want to use attribute constraints?
Have you tried this:
```
"constraints": [
   [ "rack", "UNIQUE", "spark" ]
]
```

**Comment by rodrickbrown** [comment]
@aquamatthias yes that's what I end up doing and it is now working. This issue can now be closed.

**Comment by aquamatthias** [comment]
@rodrickbrown great.
